### ISSUE TITLE
federation: incorrect use of the `ap:url` field

### ISSUE THREAD

**Issue by Johann150** [friendica/friendica#12753]
- [x] I have searched open and closed issues for duplicates

### Bug Description

The `https://www.w3.org/ns/activitystreams#url` (aka `ap:url`) property in the ActivityPub representation of `Note`s is sometimes used against the defined semantics of said field. From the definition:
> Identifies one or more links to representations *of the object*.
-- https://www.w3.org/TR/activitystreams-vocabulary/#dfn-url (emphasis mine)

I understand this to mean that the `xsd:anyURI` that is contained in `ap:url` field should lead to a representation of the object itself (i.e. the `Note`), not something it is quoting or talking about.

### Steps to Reproduce

`curl -sH 'accept: application/activity+json' 'https://social.yl.ms/objects/e18176ef85c3076f931c1e0b66aa42b328cb24df' | python3 -m json.tool`

Actual Result:

- The `ap:url` field leads to an article the `Note` is talking about.

<details><summary>view full JSON-LD example</summary>

```json
{
    "@context": [
        "https://www.w3.org/ns/activitystreams",
        "https://w3id.org/security/v1",
        {
            "vcard": "http://www.w3.org/2006/vcard/ns#",
            "dfrn": "http://purl.org/macgirvin/dfrn/1.0/",
            "diaspora": "https://diasporafoundation.org/ns/",
            "litepub": "http://litepub.social/ns#",
            "toot": "http://joinmastodon.org/ns#",
            "featured": {
                "@id": "toot:featured",
                "@type": "@id"
            },
            "schema": "http://schema.org#",
            "manuallyApprovesFollowers": "as:manuallyApprovesFollowers",
            "sensitive": "as:sensitive",
            "Hashtag": "as:Hashtag",
            "quoteUrl": "as:quoteUrl",
            "conversation": "ostatus:conversation",
            "directMessage": "litepub:directMessage",
            "discoverable": "toot:discoverable",
            "PropertyValue": "schema:PropertyValue",
            "value": "schema:value"
        }
    ],
    "id": "https://social.yl.ms/objects/e18176ef85c3076f931c1e0b66aa42b328cb24df",
    "type": "Note",
    "summary": "",
    "inReplyTo": null,
    "diaspora:guid": "e18176ef85c3076f931c1e0b66aa42b328cb24df",
    "published": "2021-09-01T10:03:18Z",
    "url": "http://www.piratenpartei-duisburg.de/bundestagswahlen-2021-und-der-hartz-iv-regelsatz/",
    "attributedTo": "https://social.yl.ms/profile/piraten_news",
    "sensitive": false,
    "content": "Laut Medienberichten soll der Hartz IV Regelsatz 2022 um genau 3,00 \u00e2\u201a\u00ac angehoben werden. Diese minimale Erh\u00c3\u00b6hung ist unserer Meinung nach eine absolute Verh\u00c3\u00b6hnung der Menschen, die von diesen Zahlungen irgendwie \u00c3\u0153berleben m\u00c3\u00bcssen. <br><a href=\"http://www.piratenpartei-duisburg.de/bundestagswahlen-2021-und-der-hartz-iv-regelsatz/\" target=\"_blank\" rel=\"noopener noreferrer\">Bundestagswahlen 2021 und der Hartz IV Regelsatz.</a><br>",
    "contentMap": {
        "de": "Laut Medienberichten soll der Hartz IV Regelsatz 2022 um genau 3,00 \u00e2\u201a\u00ac angehoben werden. Diese minimale Erh\u00c3\u00b6hung ist unserer Meinung nach eine absolute Verh\u00c3\u00b6hnung der Menschen, die von diesen Zahlungen irgendwie \u00c3\u0153berleben m\u00c3\u00bcssen. <br><a href=\"http://www.piratenpartei-duisburg.de/bundestagswahlen-2021-und-der-hartz-iv-regelsatz/\" target=\"_blank\" rel=\"noopener noreferrer\">Bundestagswahlen 2021 und der Hartz IV Regelsatz.</a>"
    },
    "source": {
        "content": "Laut Medienberichten soll der Hartz IV Regelsatz 2022 um genau 3,00 \u00e2\u201a\u00ac angehoben werden. Diese minimale Erh\u00c3\u00b6hung ist unserer Meinung nach eine absolute Verh\u00c3\u00b6hnung der Menschen, die von diesen Zahlungen irgendwie \u00c3\u0153berleben m\u00c3\u00bcssen.\n[attachment type='link' url='http://www.piratenpartei-duisburg.de/bundestagswahlen-2021-und-der-hartz-iv-regelsatz/' title='Bundestagswahlen 2021 und der Hartz IV Regelsatz.' publisher_name='Piratenpartei Duisburg' publisher_url='http://www.piratenpartei-duisburg.de' author_name='Beckmann Reinhard' author_url='http://www.piratenpartei-duisburg.de/author/kreissekretaer-kv-duisburg/'][/attachment]",
        "mediaType": "text/bbcode"
    },
    "attachment": [],
    "tag": [],
    "generator": {
        "type": "Application",
        "name": "Feed"
    },
    "to": [
        "https://www.w3.org/ns/activitystreams#Public"
    ],
    "cc": [
        "https://social.yl.ms/followers/piraten_news"
    ]
}
```
</details>

Expected Result:

- The `ap:url` field leads to a "human readable" representation of the object, e.g. the friendica frontend.

### Platform Info

Friendica Version: e.g. `2023.03-dev-1513` (`social.yl.ms` according to nodeinfo)

**Comment by Johann150** [comment]
I noticed this when I tried to add a unique index on the `url` field and noticed that there were duplicates. While my assumption that these should be unique may not be correct, it nevertheless made me notice this strange usage of `ap:url`.

There are more examples that you can probably take from https://akkoma.dev/FoundKeyGang/FoundKey/issues/331#issuecomment-7144 , I believe many of the links from there are also from Friendica instances.

**Comment by MrPetovan** [comment]
Thank you for your report!

**Comment by MrPetovan** [comment]
After review, the Friendica post in question is a news feed item, which means the provided `ap:url` as the article URL is an accurate representation of the object. We only do this in particular cases, a regular Friendica account posting a link to the same article wouldn't have the article URL in this property.

**Comment by Johann150** [comment]
I still think this is wrong for several reasons:
1. The "news feed item" and the Friendica post are different objects. Thus a link to "the object" should link to a representation of the `Note`.
2. If I see this news post and a "normal" Friendica post, there is nothing that really tells me that this is a "news feed item". It just looks like an ordinary `Note`. For me there is no real way to tell if this post was made by a friendica user that happened to input/copy the same text or if it was created from a news feed.
<details><summary>exemplary "normal" post</summary>

```
{
    "@context": [
        "https://www.w3.org/ns/activitystreams",
        "https://w3id.org/security/v1",
        {
            "vcard": "http://www.w3.org/2006/vcard/ns#",
            "dfrn": "http://purl.org/macgirvin/dfrn/1.0/",
            "diaspora": "https://diasporafoundation.org/ns/",
            "litepub": "http://litepub.social/ns#",
            "toot": "http://joinmastodon.org/ns#",
            "featured": {
                "@id": "toot:featured",
                "@type": "@id"
            },
            "schema": "http://schema.org#",
            "manuallyApprovesFollowers": "as:manuallyApprovesFollowers",
            "sensitive": "as:sensitive",
            "Hashtag": "as:Hashtag",
            "quoteUrl": "as:quoteUrl",
            "conversation": "ostatus:conversation",
            "directMessage": "litepub:directMessage",
            "discoverable": "toot:discoverable",
            "PropertyValue": "schema:PropertyValue",
            "value": "schema:value"
        }
    ],
    "id": "https://social.yl.ms/objects/e18176ef-1063-d49f-30ab-749363330842",
    "type": "Note",
    "summary": "",
    "inReplyTo": null,
    "diaspora:guid": "e18176ef-1063-d49f-30ab-749363330842",
    "published": "2023-01-28T04:06:08Z",
    "url": "https://social.yl.ms/display/e18176ef-1063-d49f-30ab-749363330842",
    "attributedTo": "https://social.yl.ms/profile/utzer",
    "sensitive": false,
    "context": "https://social.yl.ms/objects/e18176ef-1063-d49f-30ab-749363330842#context",
    "conversation": "https://social.yl.ms/objects/e18176ef-1063-d49f-30ab-749363330842#context",
    "content": "Voll bescheuert wenn man bald in nen tollen Urlaub startet und einer der Gedanken ist, dass man die <a href=\"https://social.yl.ms/search?tag=Katzen\" class=\"mention hashtag\" rel=\"tag\">#<span>Katzen</span></a> vermissen wird. \ud83d\ude15",
    "contentMap": {
        "en-us": "Voll bescheuert wenn man bald in nen tollen Urlaub startet und einer der Gedanken ist, dass man die #<a href=\"https://social.yl.ms/search?tag=Katzen\">Katzen</a> vermissen wird. \ud83d\ude15"
    },
    "source": {
        "content": "Voll bescheuert wenn man bald in nen tollen Urlaub startet und einer der Gedanken ist, dass man die #[url=https://social.yl.ms/search?tag=Katzen]Katzen[/url] vermissen wird. :-/",
        "mediaType": "text/bbcode"
    },
    "attachment": [],
    "tag": [
        {
            "type": "Hashtag",
            "href": "https://social.yl.ms/search?tag=katzen",
            "name": "#katzen"
        }
    ],
    "generator": {
        "type": "Application",
        "name": "Fediverse"
    },
    "to": [
        "https://www.w3.org/ns/activitystreams#Public"
    ],
    "cc": [
        "https://social.yl.ms/followers/utzer"
    ]
}
```
</details>

3. From what you said and from the example above I deduce that you *sometimes* use the `url` how I think it should be used, and sometimes not. This basically means the field can *sometimes* have different semantics, which basically ignores the whole reason of using ActivityPub properties with predefined semantics.

**Comment by MrPetovan** [comment]
Thank you for the elaboration! Unfortunately I can't comment further because I wasn't the one implementing the original news feed item logic from which this behavior in ActivityPub messages derives.

**Comment by annando** [comment]
This is a really old bug (older than the AP protocol) and for several years I complete forgot fixing it.

**Comment by trwnh** [comment]
if you are sharing a link to an article that exists elsewhere outside of ActivityPub you would probably want to use a Link

i can't comment further because i don't know enough about what a "news feed item" looks and works like in Friendica -- is it intended to be "alice shared a news article" and you have a preview? or is it intended to be "alice made a post" and you have a link inside that post or possibly attached to that post?

it could look anything like the following

- Announce Link (semantically closest to what's going on but not supported elsewhere currently)
- Create Link (eh, worse than Announce imo, less semantic while also being unsupported)
- Create Note with Link in `attachment` (technically could work semantically, but a lot of impls would have to change how they handle attachments to account for non-media)
- Create Note with link in `content` then Link in `tag` (this is probably best for compatibility)

my guess is you would maybe want to go with the Link in tag unless you think any of the others are more appropriate and want to convince everyone else. you could make Link tags work for either semantic use-case i suppose.

so basically

```yaml
id: example.com/activities/1
to: alice/followers
actor: alice
type: Create
object:
  id: example.com/posts/1
  url: example.com/~alice/some-post
  type: Note
  attributedTo: alice
  audience: Public
  content: <a.innerText>
  tag:
    - type: Link
      href: <a.href>
      name: <a.innerText>
```
